pkg/pullrequest/scm_test.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2019 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pullrequest
+
+import (
+	"reflect"
+	"testing"
+
+	"go.uber.org/zap"
+)
+
+func TestNewSCMHandler(t *testing.T) {
+	tests := []struct {
+		name     string
+		raw      string
+		wantRepo string
+		wantNum  int
+		wantErr  bool
+	}{
+		{
+			name:     "github",
+			raw:      "https://github.com/foo/bar/pull/1",
+			wantRepo: "foo/bar",
+			wantNum:  1,
+			wantErr:  false,
+		},
+		{
+			name:     "custom github",
+			raw:      "https://github.tekton.dev/foo/baz/pull/2",
+			wantRepo: "foo/baz",
+			wantNum:  2,
+			wantErr:  false,
+		},
+		{
+			name:     "gitlab",
+			raw:      "https://gitlab.com/foo/bar/merge_requests/3",
+			wantRepo: "foo/bar",
+			wantNum:  3,
+			wantErr:  false,
+		},
+		{
+			name:     "gitlab multi-level",
+			raw:      "https://gitlab.com/foo/bar/baz/merge_requests/3",
+			wantRepo: "foo/bar/baz",
+			wantNum:  3,
+			wantErr:  false,
+		},
+		{
+			name:    "unsupported",
+			raw:     "https://unsupported.com/foo/baz/merge_requests/3",
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			logger := &zap.SugaredLogger{}
+			got, err := NewSCMHandler(logger, tt.raw, "", "")
+			if err != nil {
+				if !tt.wantErr {
+					t.Errorf("NewSCMHandler() error = %v, wantErr %v", err, tt.wantErr)
+				}
+				return
+			}
+			if !reflect.DeepEqual(got.prNum, tt.wantNum) {
+				t.Errorf("NewSCMHandler() = %v, want %v", got, tt.wantNum)
+			}
+			if !reflect.DeepEqual(got.repo, tt.wantRepo) {
+				t.Errorf("NewSCMHandler() = %v, want %v", got, tt.wantRepo)
+			}
+		})
+	}
+}
+
+func Test_guessProvider(t *testing.T) {
+	tests := []struct {
+		name    string
+		url     string
+		want    string
+		wantErr bool
+	}{
+		{
+			name: "github",
+			url:  "https://github.com/foo/bar",
```

Comment:
For completeness, let's add test cases that might be error prone - e.g. `github.com/gitlab/foo`, `gitlab.com/github/foo`

Reply:
Nice catch, done!